src/cc_ring_array.c

```diff
@@ -142,11 +133,28 @@ ring_array_pop(void *elem, struct ring_array *arr)
     return CC_OK;
 }
 
+bool
+ring_array_empty(struct ring_array *arr)
+{
+    /* take snapshot of wpos, since another thread might be pushing */
+    uint32_t wpos = __atomic_load_n(&(arr->wpos), __ATOMIC_RELAXED);
+    return ring_array_nelem(arr->rpos, wpos, arr->cap) == 0;
+}
+
+void
+ring_array_flush(struct ring_array *arr)
+{
+    uint32_t wpos = __atomic_load_n(&(arr->wpos), __ATOMIC_RELAXED);
```

Comment:
[observation] It seems a little odd to me that all of these are `__ATOMIC_RELAXED`. I would have expected at least some to be `__ATOMIC_SEQ_CST` or `__ATOMIC_ACQ_REL` (basically enforcing _some_ form of barrier).

I don't have any experience with these in production implementations, so I don't have a suggestion, nor do I think this is wrong. If you know why this is done in this way, I'd appreciate an explanation so I can learn more about using atomics correctly.

Reply:
basically, with the producer/consumer model that the ring_array is intended for (which is used in pelikan for server/worker thread connection handoff, where server is the producer and worker is the consumer), we use a "pessimistic" view for rpos and wpos for the producer thread and the consumer thread respectively. what I mean by this is that since we cache wpos before moving rpos for the consumer thread, we know that we can at a minimum read until the cached wpos value. although wpos may have advanced since we cached it, we still only read as far as the cached wpos index. likewise, producer thread will only write as far as a cached rpos index will allow. since both threads are conservative about how much they produce/consume, there is no risk of over reading/over writing, no matter what order the api calls come in, which allows us to use relaxed atomicity ordering.

Reply:
Wow, ok, thank you for the explanation, that makes sense 👍